cirq-core/cirq/transformers/analytical_decompositions/quantum_shannon_decomposition.py

```diff
@@ -0,0 +1,247 @@
+
+    # Add ops from decomposition of multiplexed u1/u2 part
+    _msb_demuxer(qubits, u1, u2, ops)
+
+    # All operations are returned in order
+    return ops
+
+
+def _single_qubit_decomposition(qubit, U, ops=None):
+    """Decomposes single-qubit gate, and returns list of operations, keeping phase invariant.
+    Intended to also append these operations to existing operation list
+
+    Args:
+        qubit: Qubit on which to apply operations
+        U: (2 x 2) Numpy array for unitary representing 1-qubit gate to be decomposed
+        ops: List of new existing operations on which to append new operations, whenever
+                a recusive call is made
+             If 'None' is given, a new list is instantiated
+
+    Returns:
+        Initial operations list with new 3 operations (rz,ry,ZPowGate) added
+    """
+    if ops is None:
+        ops = []  # Declare an empty list if no previous operations
+
+    # Perform native ZYZ decomposition
+    phi_0, phi_1, phi_2 = cirq.deconstruct_single_qubit_matrix_into_angles(U)
+
+    # Determine global phase picked up
+    phase = np.angle(U[0, 0] / (np.exp(-1j * (phi_0) / 2) * np.cos(phi_1 / 2)))
+
+    # Append first two operations operations
+    ops.append(cirq.rz(phi_0).on(qubit))
+    ops.append(cirq.ry(phi_1).on(qubit))
+
+    # Append third operation with global phase added
+    ops.append(cirq.ZPowGate(exponent=phi_2 / np.pi, global_shift=phase / phi_2).on(qubit))
+    return ops
+
+
+def _msb_demuxer(demux_qubits: list, u1: np.ndarray, u2: np.ndarray, ops=None):
+    """Demultiplexes a unitary matrix that is multiplexed in in its most-significant-qubit
+    Decomposition structure:
+      [ u1 , 0  ]  =  [ V , 0 ][ D , 0  ][ W , 0 ]
+      [ 0  , u2 ]     [ 0 , V ][ 0 , D* ][ 0 , W ]
+
+     Gives: ( u1 )( u2* ) = ( V )( D^2 )( V* )
+       and:  W = ( D )( V* )( u2 )
+
+
+    Args:
+        demux_qubits: Subset of total qubits involved in this unitary gate
+        u1: Upper-left quadrant of total unitary to be decomposed (see diagram)
+        u2: Lower-right quadrant of total unitary to be decomposed (see diagram)
+        ops: List of new existing operations on which to append new operations, whenever
+                a recusive call is made
+             If 'None' is given, a new list is instantiated
+
+    Calls:
+        1. quantum_shannon_decomposition
+        2. _multiplexed_cossin
+        3. quantum_shannon_decomposition
+
+    Returns: List of 2-qubit and 1-qubit operations
+    """
+    # Perform a diagonalization to find values
+    U = u1 @ u2.T.conjugate()
+    dsquared, V = np.linalg.eig(U)
+    d = np.sqrt(dsquared)
+    D = np.diag(d)
+    W = D @ V.T.conjugate() @ u2
+    if ops is None:
+        ops = []  # Declare an empty list if no previous operations
+
+    # Last term is given by ( I ⊗ W ), demultiplexed
+    # Remove most-significant (demuxed) control-qubit
+    # Add operations for QSD on W
+    quantum_shannon_decomposition(demux_qubits[1:], W, ops)
```

Comment:
For all methods, I'd recommend you return a `cirq.OP_TREE` and then instead of constructing a list of operations and returning it you can directly yield from the recursive function calls. i.e.
```suggestion
    yield from quantum_shannon_decomposition(demux_qubits[1:], W, ops)
```

Reply:
I like this suggestion. I've implemented all the methods with yield, and introduced a top-level quantum_shannon_decomposition that consolidates into a a list

Reply:
Imlemented yield from in this place. Changed some other things too